Legger til bygging av typescript-komponentet og koverterer knapper (Resolves #70)

etse opened the pull request:
La til et oppsett for å bygge TypeScript-komponenter, og konverterte Knapper og Alertstriper til typescript. I den forbindelse testet jeg om den bygde koden fungerte i FO-portefølje, hvor jeg fikk noen byggefeil. Årsaken til dette var at portefølje hadde feil i koden sin (brukte knappene feil), og de eksisterende definisjonstypene tillot dette.

Vil anbefale at man gjerne testet dette godt før man merger inn, spesielt med tanke på knapper i JS-applikasjoner.

Når denne blir merget skal det ligge til rette for å enkelt kunne konvertere flere komponenter til typescript fremover.

Resolves #70 
Resolves #155 
Resolves #156 
Resolves #157 
Resolves #166
Resolves #168
Resolves #170 
Resolves #171

etse committed:
Legger til bygging av typescript-komponentet og koverterr knapper

etse committed:
Konverterer AlertStriper til typescript

etse committed:
Endrer til string-union-type i stede for enum (mindre breacking change)

etse committed:
Fikser lint-feil

etse committed:
Merge remote-tracking branch 'origin/master' into typescript

# Conflicts:
#	packages/node_modules/nav-frontend-alertstriper/src/navansatt-alertstripe.js

nutgaard commented on packages/node_modules/nav-frontend-alertstriper/package.json:
```diff
@@ -2,7 +2,7 @@
   "version": "0.2.17",
   "main": "lib/alertstripe.js",
-  "jsnext:main": "src/alertstripe.js",
+  "types": "lib/alertstripe.d.ts",
```
Vi vil vel fortsatt ha `jsnext:main` eller?

etse replied:
det finnes ikke noen jsnext lenger, er enten es5 eller TS. Og blir feil å la jsnext:main peke til noen av delene. Med andre ord, "test-appen" burde enten bruke de bygde filene eller gå selv konverteres til typescript.

nutgaard replied:
True true. 😛

nutgaard commented on packages/node_modules/nav-frontend-alertstriper/src/advarsel-alertstripe.tsx:
```diff
@@ -0,0 +1,10 @@
+import * as React from 'react';
+import AlertStripe, {AlertStripeProps} from './alertstripe';
+
+function AlertStripeAdvarsel(props: AlertStripeProps) {
```
`function AlertStripeAdvarsel(props: AlertStripeProps): JSXElement {}` ? 
Da slipper du vel casting til `any` nedenfor også?

etse replied:
Ville trodd type-inferensen skulle hjulpet her. Men tydeligvis ikke helt optimal. Og fikk ikke løsningen du foreslo.

Prøvde litt ulike ting, fant ikke en optimal løsning, men kan caste til noe mer fornuftig enn any.

nutgaard replied:
Hvis vi hadde brukt `const AlertStripeAdvarsel = function(){}` syntaxen så ville det fungert. Men har sett det er par ganger nå at typescript ikke helt forstår seg på typene når det er vanlige funksjoner.

Ett alternativ er å skrive om til `React.PureComponent` og legge til `static propTypes = {};` i klassen. (Om vi kjører med ny nok versjon av babel vel og merke. :P)

etse replied:
Å bruke const vil nok fungere fint her ja. Kan fint endre til det om det er ønskelig. Eneste problemet er at det kan føre til rare feil med tanke på hoisting (vært borti at det skjedde i portefølje). Tror årsaken egentlig var sirkulære avhengigheter hvor omskriving  til const gjør at det ikke blir hoista og man da ble avhengig av rekkefølgen til ble resolvet i. Så tror ikke det burde være noe issue her.

nutgaard replied:
Jeg vil egentlig ikke ha `const`-syntaksen, men eneste grunnen til at jeg ikke liker den er pga at jeg synes det ser stygt ut. :P

nutgaard commented on packages/node_modules/nav-frontend-knapper/src/knapp.tsx:
```diff
@@ -0,0 +1,19 @@
+import {CustomHTMLButtonAttributes} from "./CustomHTMLButtonAttributes";
+
+export interface KnappProps extends CustomHTMLButtonAttributes {
+    htmlType?: 'submit' | 'button' | 'reset',
```
Hva får vi egentlig fra `CustomHTMLButtonAttributes` som vi ikke får fra `ButtonHTMLAttributes`?

etse replied:
type er av typen string i `ButtonHTMLAttributes`. Dette er ikke optimalt for oss, siden vi bruker den som en "enum" (valgte å ikke gjøre til enum siden det e breaking-change, den er derfor en string-union-type). 

Når man extender et interface har man ikke mulighet til å override typen til felter. (og når jeg leste issues andre har om dette virker det som at folk ikke ønsker å fikse det heller da det egentlig er en codesmell).

Skulle gjerne renamet `type` til `knappType` og gjort `htmlType` til `type`. Da slipper jeg å extende min egen custom-ting. Men det var en såpas stor breaking change at jeg valgte å la være.

nutgaard replied:
Det har vært ett issue for flere opp igjennom. Var lite gjennomtenkt å bruke `type`-navnet her til å starte med. :( 

Men spørsmålet er om vi burde rydde opp i dette først som sist, e.g. innføre den endringen du nevner i siste avsnitt. @erlendev @frodehansen2 @helgefredheim @pederpus Tanker?

nutgaard commented on packages/node_modules/nav-frontend-knapper/src/knapp.tsx:
```diff
@@ -0,0 +1,19 @@
+    inaktivKlasseVedDisabled?: boolean
+}
+
+class Knapp extends React.Component<KnappProps> { // eslint-disable-line react/prefer-stateless-function
```
Generelt angående propsTypes; må disse nå forvaltes som både en typescript- type og propTypes? Eller finnes det noen automatikk her?

etse replied:
Ikke finnet noe automatisk. På sikt hadde det beste vært å bare slette propTypes om majoriteten av prosjektene brukte TS. Men enn så lenge er det såpass mye prosjekter med ren JS at jeg tror å forvalte begge deler gir stor verdi.

nutgaard replied:
Fant noe som kanskje kunne løse noe av dette på interwebben i går. Men virker veldig uklart og litt shady. Så greieste er vel at vi bare forvalter begge frem til vi ser hvordan det går med adopsjonen av TypeScript her på huset.

nutgaard reviewed (changes requested):
🥇 Knallbra.
Har en par spørsmål though. :)

etse committed:
Ikke caste til any, men mer fornuftige typer i stede

etse committed:
Endre syntax på stateless component

etse committed:
Merge remote-tracking branch 'origin/master' into typescript

etse committed:
Legger til coffee-script som eksplisitt dependency grunnt issue siden 1.12.8 ikke fungerer

Commit:
Merge remote-tracking branch 'origin/master' into typescript

# Conflicts:
#	packages/node_modules/nav-frontend-alertstriper-style/package.json
#	packages/node_modules/nav-frontend-alertstriper/package.json
#	packages/node_modules/nav-frontend-hjelpetekst/package.json
#	packages/node_modules/nav-frontend-spinner/package.json
#	packages/node_modules/nav-frontend-stegindikator-style/package.json

Commit:
Fikser import av prop-types

Commit:
Konverterer tekstområde til typescript

Resolves #170

Commit:
Konverterer typografi, panel og snakkeboble til typescript

Snakkeboble bruker typografi og panel så de måtte oppdateres sammen.

Resolves #166
Resolves #168
Resolves #171

frodehansen2 committed:
Typescript Chevron

frodehansen2 committed:
Stateless

etse committed:
Merge pull request #189 from navikt/issue/157-chevron-ts

Issue/157 Chevron typescript

etse committed:
Merge pull request #186 from navikt/fix/168-konvertere-snakkeboble-til-ts

Konverterer typografi, panel og snakkeboble til typescript

etse committed:
Merge pull request #180 from navikt/fix/170-konvertere-tekstomrade-til-ts

Konverterer tekstområde til typescript
